src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs

```diff
@@ -14,10 +14,11 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
     internal partial class Http2Stream : IHttp2StreamIdFeature,
                                          IHttpMinRequestBodyDataRateFeature,
                                          IHttpResetFeature,
-                                         IHttpResponseTrailersFeature
-
+                                         IHttpResponseTrailersFeature,
+                                         IPersistentStateFeature
     {
         private IHeaderDictionary? _userTrailers;
+        private IDictionary<object, object?>? _persistentState;
```

Comment:
Customize HTTP/2 so that state is preserved per-HTTP/2 stream. We don't want to share a dictionary across all streams on a connection.

Reply:
Can we add a comment above the `_persistentState` declarations to say this shouldn't ever be reset since it's used for pooling?

Looking at this got me looking at whether `_userTrailers` gets reset. It doesn't look like it does.

Reply:
> Looking at this got me looking at whether _userTrailers gets reset. It doesn't look like it does.

I tested it. It is set to ResponseTrailers which is then cleared in HttpProtocol.